File: src/formats/fb2/parser.test.ts
import { describe, it, expect } from 'vitest';
import { parseFb2Text, parseFb2Buffer, extractFb2FromZip } from './parser.js';
import { FB2_SAMPLE, FB2_CP1251_SAMPLE, makeFb2Zip, zipFileNames } from '../test-utils.js';
import { ParseError } from '../../utils/errors.js';
import { joinAuthors } from '../model.js';
import { detectFormat } from '../index.js';

describe('FB2 parser', () => {
  it('parses metadata from title-info', () => {
    const { metadata } = parseFb2Text(FB2_SAMPLE, '/tmp/book.fb2', 'book.fb2');
    expect(metadata.title).toBe('Test Book');
    expect(metadata.authors).toHaveLength(2);
    expect(joinAuthors(metadata.authors)).toBe('Doe John, Nick');
    expect(metadata.genres).toEqual(['sf', 'adventure']);
    expect(metadata.annotation).toContain('important');
    expect(metadata.lang).toBe('en');
    expect(metadata.series).toEqual({ name: 'The Series', number: 2 });
    expect(metadata.coverKey).toBe('cover.jpg');
    expect(metadata.publisher).toBe('Example Press');
    expect(metadata.year).toBe(2020);
    expect(metadata.isbn).toBe('978-3-16-148410-0');
  });

  it('falls back to filename for missing title', () => {
    const xml = `<?xml version="1.0"?><FictionBook><description/></FictionBook>`;
    const { metadata } = parseFb2Text(xml, '/tmp/no-title.fb2', 'no-title.fb2');
    expect(metadata.title).toBe('no-title');
  });

  it('parses content blocks preserving structure', () => {
    const { state } = parseFb2Text(FB2_SAMPLE, '/tmp/book.fb2', 'book.fb2');
    const types = state.blocks.map((b) => b.type);
    expect(types).toContain('heading');
    expect(types).toContain('paragraph');
    expect(types).toContain('list');
    expect(types).toContain('quote');
    expect(types).toContain('poem');
    expect(types).toContain('empty');
    expect(types).toContain('table');

    const paragraphs = state.blocks.filter((b) => b.type === 'paragraph');
    const first = paragraphs[0];
    expect(first).toBeDefined();
    if (first && first.type === 'paragraph') {
      const hasBold = first.children.some(
        (c) =>
          c.kind === 'bold' && c.children.some((x) => x.kind === 'text' && x.text.includes('bold')),
      );
      expect(hasBold).toBe(true);
    }

    const heading = state.blocks.find((b) => b.type === 'heading');
    expect(heading).toBeDefined();
    if (heading && heading.type === 'heading') {
      expect(heading.level).toBe(2);
    }
  });

  it('builds a table of contents', () => {
    const { state } = parseFb2Text(FB2_SAMPLE, '/tmp/book.fb2', 'book.fb2');
    const labels = state.toc.map((t) => t.label);
    expect(labels).toContain('Book Title');
    expect(labels).toContain('Chapter One');
    expect(labels).toContain('Nested Section');
    for (const entry of state.toc) {
      const block = state.blocks[entry.blockIndex];
      expect(block).toBeDefined();
      expect(block?.type).toBe('heading');
    }
  });

  it('collects binary resources', () => {
    const { resources } = parseFb2Text(FB2_SAMPLE, '/tmp/book.fb2', 'book.fb2');
    expect(resources.has('cover.jpg')).toBe(true);
    expect(resources.has('img1')).toBe(true);
    const cover = resources.get('cover.jpg')!;
    expect(Buffer.from(cover).toString('utf8')).toBe('hello world');
  });

  it('decodes windows-1251 encoded FB2', () => {
    const book = parseFb2Buffer(FB2_CP1251_SAMPLE, '/tmp/ru.fb2');
    expect(book.metadata.title).toBe('Тестовая книга');
    expect(book.metadata.authors[0]?.firstName).toBe('Иван');
  });

  it('parses fb2.zip archives', () => {
    const zip = makeFb2Zip(FB2_SAMPLE, 'books/test.fb2');
    const book = parseFb2Buffer(zip, '/tmp/archive.fb2.zip');
    expect(book.format).toBe('fb2');
    expect(book.metadata.title).toBe('Test Book');
    expect(zipFileNames(zip)).toContain('books/test.fb2');
  });

  it('extracts the inner fb2 from a zip', () => {
    const zip = makeFb2Zip(FB2_SAMPLE);
    const inner = extractFb2FromZip(zip);
    expect(inner.name).toBe('book.fb2');
    expect(inner.data.length).toBeGreaterThan(0);
  });

  it('throws ParseError for non-FB2 XML', () => {
    expect(() => parseFb2Text('<html></html>', '/tmp/x.fb2', 'x.fb2')).toThrow(ParseError);
  });

  it('throws ParseError when a zip has no fb2 file', () => {
    expect(() => parseFb2Buffer(makeFb2Zip('<html/>', 'readme.txt'), '/tmp/a.fb2.zip')).toThrow(
      ParseError,
    );
  });

  it('detects format from content and extension', () => {
    expect(detectFormat(Buffer.from(FB2_SAMPLE), 'unknown.txt')).toBe('fb2');
    expect(detectFormat(makeFb2Zip(FB2_SAMPLE), 'x.fb2.zip')).toBe('fb2');
    expect(detectFormat(Buffer.from('PK\x03\x04 data'), 'x.epub')).toBe('epub');
  });

  it('handles truncated xml gracefully without crashing', () => {
    const corrupt = '<?xml version="1.0"?><FictionBook><description><broken>';
    const book = parseFb2Text(corrupt, '/tmp/b.fb2', 'b.fb2');
    expect(book.metadata.title).toBe('b');
    expect(book.state.blocks.length).toBe(0);
  });
});

describe('FB2 parser coverage edge cases', () => {
  const fb2Doc = (body: string): string =>
    `<?xml version="1.0"?><FictionBook xmlns="http://www.gribuser.ru/xml/fictionbook/2.0"><description><title-info><book-title>T</book-title></title-info></description><body>${body}</body></FictionBook>`;

  it('falls back to filename when description is missing', () => {
    const { metadata, state } = parseFb2Text(
      '<?xml version="1.0"?><FictionBook><body><p>Hi</p></body></FictionBook>',
      '/tmp/x.fb2',
      'x.fb2',
    );
    expect(metadata.title).toBe('x');
    expect(state.blocks.map((b) => b.type)).toContain('paragraph');
  });

  it('parses middle-name-only authors and skips empty authors', () => {
    const xml = `<?xml version="1.0"?><FictionBook><description><title-info>
      <book-title>T</book-title>
      <author><middle-name>Petrovich</middle-name></author>
      <author><first-name/></author>
    </title-info></description><body/></FictionBook>`;
    const { metadata } = parseFb2Text(xml, '/tmp/a.fb2', 'a.fb2');
    expect(metadata.authors).toEqual([
      { firstName: '', lastName: '', middleName: 'Petrovich', nickname: '' },
    ]);
  });

  it('falls back to filename for an empty book-title and missing lang', () => {
    const xml = `<?xml version="1.0"?><FictionBook><description><title-info><book-title>   </book-title></title-info></description><body/></FictionBook>`;
    const { metadata } = parseFb2Text(xml, '/tmp/e.fb2', 'e.fb2');
    expect(metadata.title).toBe('e');
    expect(metadata.lang).toBeUndefined();
  });

  it('handles sequences with invalid numbers and a missing coverpage', () => {
    const xml = `<?xml version="1.0"?><FictionBook><description><title-info>
      <book-title>T</book-title>
      <sequence name="Saga" number="abc"/>
    </title-info></description><body/></FictionBook>`;
    const { metadata } = parseFb2Text(xml, '/tmp/s.fb2', 's.fb2');
    expect(metadata.series).toEqual({ name: 'Saga', number: undefined });
    expect(metadata.coverKey).toBeUndefined();
  });

  it('handles publish-info without publisher or isbn', () => {
    const xml = `<?xml version="1.0"?><FictionBook><description><title-info><book-title>T</book-title></title-info><publish-info><year>2020</year></publish-info></description><body/></FictionBook>`;
    const { metadata } = parseFb2Text(xml, '/tmp/p.fb2', 'p.fb2');
    expect(metadata.publisher).toBeUndefined();
    expect(metadata.isbn).toBeUndefined();
    expect(metadata.year).toBe(2020);
  });

  it('skips binary resources without an id', () => {
    const xml = `<?xml version="1.0"?><FictionBook><description><title-info><book-title>T</book-title></title-info></description><body/><binary>aGVsbG8=</binary></FictionBook>`;
    const { resources } = parseFb2Text(xml, '/tmp/b.fb2', 'b.fb2');
    expect(resources.size).toBe(0);
  });

  it('parses exotic content blocks', () => {
    const body = `
      <title>Plain Title</title>
      <title><p><strong>B</strong><emphasis>I</emphasis><u>U</u><s>S</s><a l:href="#x">L</a><code>C</code><br/>img</p></title>
      <section><title>Sec</title><p>X</p></section>
      <p>   </p>
      <list><li>Outer<list><li>Inner</li></list></li></list>
      <poem><title>Poem Title</title><v>Line</v><subtitle>Sub</subtitle><stanza><v>V1</v><v>V2</v></stanza></poem>
      <poem><title><p>WithP</p></title><stanza><v>X</v></stanza></poem>
      <table><tr><td>A</td><td>B</td></tr></table>
      <epigraph><p>Quote</p><text-author>Author</text-author></epigraph>
      <epigraph><poem><stanza><v>Verse</v></stanza></poem></epigraph>
      <epigraph><subtitle>Lone</subtitle></epigraph>
      <annotation><p>Note</p></annotation>
      <text-author>Me</text-author>
      <code>const x = 1;</code>
      <image l:href="#pic" alt="pic"/>
      <image/>
      <unknown><p>Inner</p></unknown>
      <unknown/>
    `;
    const { state } = parseFb2Text(fb2Doc(body), '/tmp/c.fb2', 'c.fb2');
    const types = state.blocks.map((b) => b.type);
    for (const t of [
      'heading',
      'paragraph',
      'list',
      'poem',
      'table',
      'epigraph',
      'annotation',
      'image',
      'empty',
    ]) {
      expect(types).toContain(t);
    }
    expect(state.toc.length).toBeGreaterThan(0);
    const table = state.blocks.find((b) => b.type === 'table');
    if (table && table.type === 'table') {
      expect(table.headers).toEqual([]);
      expect(table.rows).toHaveLength(1);
    }
  });

  it('throws when extracting from a zip without an fb2 file', () => {
    expect(() => extractFb2FromZip(makeFb2Zip('<html/>', 'readme.txt'))).toThrow(ParseError);
  });
});

File: crates/tabook-native/src/fb2/parser_test.rs
//! FB2 parser tests — port of `src/formats/fb2/parser.test.ts`.

use super::parser::*;

const FB2_SAMPLE: &str = r##"<?xml version="1.0" encoding="UTF-8"?>
<FictionBook xmlns="http://www.gribuser.ru/xml/fictionbook/2.0">
  <description>
    <title-info>
      <genre>sf</genre>
      <genre>adventure</genre>
      <author>
        <first-name>John</first-name>
        <last-name>Doe</last-name>
      </author>
      <author>
        <nickname>Nick</nickname>
      </author>
      <book-title>Test Book</book-title>
      <annotation><p>An <emphasis>important</emphasis> annotation.</p></annotation>
      <lang>en</lang>
      <sequence name="The Series" number="2"/>
      <coverpage><image l:href="#cover.jpg"/></coverpage>
    </title-info>
    <publish-info>
      <publisher>Example Press</publisher>
      <year>2020</year>
      <isbn>978-3-16-148410-0</isbn>
    </publish-info>
  </description>
  <body>
    <title><p>Book Title</p></title>
    <section id="ch1">
      <title><p>Chapter One</p></title>
      <p>First paragraph with <strong>bold</strong> and <emphasis>italic</emphasis> text.</p>
      <p>A <a l:href="#note1">footnote link</a> and an inline image <image l:href="#img1" alt="diagram"/>.</p>
      <list><li>Item one</li><li>Item two</li></list>
      <cite><p>To be or not to be.</p></cite>
      <poem>
        <stanza>
          <v>Line one of verse</v>
          <v>Line two of verse</v>
        </stanza>
      </poem>
      <empty-line/>
      <p>Second paragraph.</p>
      <subtitle>A Subtitle</subtitle>
      <table>
        <tr><th>Col A</th><th>Col B</th></tr>
        <tr><td>1</td><td>2</td></tr>
      </table>
      <section id="ch1.1">
        <title><p>Nested Section</p></title>
        <p>Nested content.</p>
      </section>
    </section>
  </body>
  <binary id="cover.jpg" content-type="image/jpeg">aGVsbG8gd29ybGQ=</binary>
  <binary id="img1" content-type="image/png">aW1nZGF0YQ==</binary>
</FictionBook>"##;

#[test]
fn parses_metadata() {
    let result = parse_fb2_text(FB2_SAMPLE, "/test.fb2", "test.fb2").unwrap();
    assert_eq!(result.metadata.title, "Test Book");
    assert_eq!(result.metadata.authors.len(), 2);
    assert_eq!(result.metadata.authors[0].first_name, "John");
    assert_eq!(result.metadata.authors[0].last_name, "Doe");
    assert_eq!(result.metadata.authors[1].nickname.as_deref(), Some("Nick"));
    assert_eq!(result.metadata.genres, vec!["sf", "adventure"]);
    assert_eq!(result.metadata.lang.as_deref(), Some("en"));
    assert_eq!(result.metadata.series.as_ref().unwrap().name, "The Series");
    assert_eq!(result.metadata.series.as_ref().unwrap().number, Some(2.0));
    assert_eq!(result.metadata.cover_key.as_deref(), Some("cover.jpg"));
    assert_eq!(result.metadata.publisher.as_deref(), Some("Example Press"));
    assert_eq!(result.metadata.isbn.as_deref(), Some("978-3-16-148410-0"));
    assert_eq!(result.metadata.year, Some(2020.0));
}

#[test]
fn parses_body_blocks() {
    let result = parse_fb2_text(FB2_SAMPLE, "/test.fb2", "test.fb2").unwrap();
    // The first block should be the Book Title heading
    assert!(result.blocks.len() > 5);
    let first = &result.blocks[0];
    assert_eq!(first.r#type, "heading");
    // Chapter One heading
    let ch1 = result.blocks.iter().find(|b| {
        b.r#type == "heading"
            && b.children
                .as_deref()
                .map(|c| c.iter().any(|i| i.text.as_deref() == Some("Chapter One")))
                .unwrap_or(false)
    });
    assert!(ch1.is_some());
}

#[test]
fn parses_toc() {
    let result = parse_fb2_text(FB2_SAMPLE, "/test.fb2", "test.fb2").unwrap();
    assert!(!result.toc.is_empty());
    let ch1 = result.toc.iter().find(|t| t.label == "Chapter One");
    assert!(ch1.is_some());
    assert_eq!(ch1.unwrap().level, 2);
}

#[test]
fn parses_resources() {
    let result = parse_fb2_text(FB2_SAMPLE, "/test.fb2", "test.fb2").unwrap();
    assert_eq!(result.resources.len(), 2);
    let cover = result.resources.iter().find(|r| r.key == "cover.jpg");
    assert!(cover.is_some());
    // "aGVsbG8gd29ybGQ=" decodes to "hello world"
    assert_eq!(&cover.unwrap().data[..], b"hello world");
}

#[test]
fn parses_list_block() {
    let result = parse_fb2_text(FB2_SAMPLE, "/test.fb2", "test.fb2").unwrap();
    let list = result.blocks.iter().find(|b| b.r#type == "list");
    assert!(list.is_some());
    let items = list.unwrap().items.as_ref().unwrap();
    assert_eq!(items.len(), 2);
}

#[test]
fn parses_poem_block() {
    let result = parse_fb2_text(FB2_SAMPLE, "/test.fb2", "test.fb2").unwrap();
    let poem = result.blocks.iter().find(|b| b.r#type == "poem");
    assert!(poem.is_some());
    let stanzas = poem.unwrap().stanzas.as_ref().unwrap();
    assert_eq!(stanzas.len(), 1);
    assert_eq!(stanzas[0].lines.len(), 2);
}

#[test]
fn parses_table_block() {
    let result = parse_fb2_text(FB2_SAMPLE, "/test.fb2", "test.fb2").unwrap();
    let table = result.blocks.iter().find(|b| b.r#type == "table");
    assert!(table.is_some());
    let t = table.unwrap();
    assert_eq!(t.headers.as_ref().unwrap().len(), 2);
    assert_eq!(t.rows.as_ref().unwrap().len(), 1);
}

#[test]
fn parses_quote_block() {
    let result = parse_fb2_text(FB2_SAMPLE, "/test.fb2", "test.fb2").unwrap();
    let quote = result.blocks.iter().find(|b| b.r#type == "quote");
    assert!(quote.is_some());
}

#[test]
fn metadata_only_is_faster_path() {
    let data = FB2_SAMPLE.as_bytes();
    let metadata = parse_fb2_metadata_inner(data, "/test.fb2").unwrap();
    assert_eq!(metadata.title, "Test Book");
    assert_eq!(metadata.authors.len(), 2);
}

#[test]
fn parse_fb2_buffer_from_bytes() {
    let data = FB2_SAMPLE.as_bytes();
    let book = parse_fb2_buffer_inner(data, "/test.fb2").unwrap();
    assert_eq!(book.format, "fb2");
    assert_eq!(book.metadata.title, "Test Book");
    assert!(!book.content.is_empty());
    assert_eq!(book.resources.len(), 2);
}

#[test]
fn parse_fb2_zip() {
    use std::io::Write;
    let mut buf = Vec::new();
    {
        let mut zip = zip::ZipWriter::new(std::io::Cursor::new(&mut buf));
        let opts = zip::write::SimpleFileOptions::default();
        zip.start_file("book.fb2", opts).unwrap();
        zip.write_all(FB2_SAMPLE.as_bytes()).unwrap();
        zip.finish().unwrap();
    }
    let book = parse_fb2_buffer_inner(&buf, "/test.fb2.zip").unwrap();
    assert_eq!(book.format, "fb2");
    assert_eq!(book.metadata.title, "Test Book");
}

#[test]
fn rejects_non_fb2() {
    let result = parse_fb2_buffer_inner(b"<html>not fb2</html>", "/test.xml");
    assert!(result.is_err());
}

#[test]
fn zip_metadata_fallback_uses_basename() {
    // Regression: the metadata-only zip path used to strip the extension off
    // the FULL archive path, so a zip with the fb2 in a subdirectory got
    // "subdir/book" as the fallback title for books without <book-title>.
    // The full-parse path already used the entry basename (parity:
    // src/parity/fb2.parity.test.ts).
    use std::io::Write;
    let minimal = r##"<?xml version="1.0" encoding="UTF-8"?>
<FictionBook xmlns="http://www.gribuser.ru/xml/fictionbook/2.0">
  <description><title-info>
    <author><first-name>Ann</first-name><last-name>Lee</last-name></author>
  </title-info></description>
  <body><section><title><p>Chapter</p></title><p>Body text.</p></section></body>
</FictionBook>"##;
    let mut buf = Vec::new();
    {
        let mut zip = zip::ZipWriter::new(std::io::Cursor::new(&mut buf));
        let opts = zip::write::SimpleFileOptions::default();
        zip.start_file("subdir/book.fb2", opts).unwrap();
        zip.write_all(minimal.as_bytes()).unwrap();
        zip.finish().unwrap();
    }
    let metadata = parse_fb2_metadata_inner(&buf, "/test.fb2.zip").unwrap();
    assert_eq!(metadata.title, "book");
    let book = parse_fb2_buffer_inner(&buf, "/test.fb2.zip").unwrap();
    assert_eq!(book.metadata.title, "book");
}
